src/libraries/System.Text.Json/tests/Serialization/CustomConverterTests/CustomConverterTests.HandleNull.cs

```diff
@@ -0,0 +1,484 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Buffers;
+using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+using Xunit;
+
+namespace System.Text.Json.Serialization.Tests
+{
+    public static partial class CustomConverterTests
+    {
+        [Fact]
+        public static void ValueTypeConverter_NoOverride()
+        {
+            // Baseline
+            Assert.Throws<JsonException>(() => JsonSerializer.Deserialize<int>("null"));
+
+            // Per null handling default value for value types (true), converter handles null.
+            var options = new JsonSerializerOptions();
+            options.Converters.Add(new Int32NullConverter_SpecialCaseNull());
+
+            Assert.Equal(-1, JsonSerializer.Deserialize<int>("null", options));
+        }
+
+        private class Int32NullConverter_SpecialCaseNull : JsonConverter<int>
+        {
+            public override int Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
+            {
+                if (reader.TokenType == JsonTokenType.Null)
+                {
+                    return -1;
+                }
+
+                throw new JsonException();
+            }
+
+            public override void Write(Utf8JsonWriter writer, int value, JsonSerializerOptions options)
+            {
+                throw new NotImplementedException();
+            }
+        }
+
+        [Fact]
+        public static void ValueTypeConverter_OptOut()
+        {
+            // Per null handling opt-out, serializer handles null.
+            var options = new JsonSerializerOptions();
+            options.Converters.Add(new Int32NullConverter_OptOut());
+
+            Assert.Equal(0, JsonSerializer.Deserialize<int>("null", options));
```

Comment:
Shouldn't this throw? Serializer would normally throw for non-nullable types.

Reply:
The exception is thrown from the converter. Most of the built-in value converters will throw if the token type is not `Number` or `String`.

Here, the converter to be used is the custom one, but it has opted out of handling `null`. Rather than passing it to the built-in Int32 converter (we have no reference to it at this point), the serializer sets the `default` value for the type:

https://github.com/dotnet/runtime/blob/b0329b19adf831cf81e9488c3de3547472784b2f/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonConverterOfT.cs#L115

https://github.com/dotnet/runtime/blob/b0329b19adf831cf81e9488c3de3547472784b2f/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonConverterOfT.cs#L158

https://github.com/dotnet/runtime/blob/b0329b19adf831cf81e9488c3de3547472784b2f/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonConverterOfT.cs#L174

https://github.com/dotnet/runtime/blob/b0329b19adf831cf81e9488c3de3547472784b2f/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonPropertyInfoOfT.cs#L167

We could add another check and only set the `default` when the type can be null (otherwise throw JsonException) if desired, but for now this test reflects our current behavior.

Reply:
I think we should discuss this a bit more.

Before this PR a value-type custom converter gets passed null which it would normally throw when it calls reader.GetInt32() etc but the converter can decide what to do (throw or do something with the null - like we do for `JsonElement` and `Nullable<T>`).

With this PR if the converter overrides HandleNull and returns `false` then the default value is assigned and the converter is not called. I was assuming the serializer would throw `JsonException` in this case, but after thinking about it, the expected semantics are somewhat ambiguous.

However I do think throwing for value types is the "less surprising" expectation than returning the default.

The references to `value = default` above point to code that was written only to handle reference types, since we made value types handle null and reference types never handle null (so `default` would always be null).

FWIW I wouldn't expect a value-type converter to override HandleNull and return `false`, so this is an edge case anyway. We added this feature \ PR for reference-type converters.

Reply:
Per offline discussion, we should throw JsonException when `null` is an input for types that can't be null & the converter has opted out of handling null values. I'll add a commit adding this.